compiler/rustc_query_system/src/dep_graph/serialized.rs

```diff
@@ -34,138 +39,148 @@ rustc_index::newtype_index! {
 }
 
 /// Data for use when recompiling the **current crate**.
-#[derive(Debug)]
+///
+/// The DepGraph is backed on-disk and read on-demand through a Mmap.
+/// The file layout is as follows.
+///
+/// The DepGraph starts with the version header, handled by rustc_incremental.
+/// It is followed by the concatenation of all node dependency information as:
+/// - the query result fingerprint (size_of<Fingerprint> bytes)
+/// - the number of dependency edges (4 bytes)
+/// - the dependencies indices (array of 4-byte integers)
+///
+/// Finding this information inside the file is handled by a "glossary" written at the end.
+/// This glossary is an array of `(DepNode, u32)` pairs.  This array is used to make the
+/// correspondence between the `SerializedDepNodeIndex` (ie. the index into this array),
+/// and the `DepNode`.  The `u32` is the position of the dependency information (Fingerprint +
+/// array of dependencies) inside the file.  The glossary array is directly mmapped into `nodes`.
+///
+/// The file finished with two `u64`, which are the number of entries in the glossary
+/// and its position in the file.
+///
+/// Graphically, we have:
+///          beginning of nodes                        beginning of glossary ---------------+
+///          v                                         v                                    |
+/// --------------------------------------------------------------------------------------------
+/// | HEADER | ... | Fingerprint | Length | Deps | ... | ... | DepNode | u32 | ... | u64 | u64 |
+/// --------------------------------------------------------------------------------------------
+///                ^                                                      |         node
+///                start for node i --------------------------------------+         count
+///
+/// In order to recover an index from a DepNode, we populate a hash-map in `index`.
 pub struct SerializedDepGraph<K: DepKind> {
-    /// The set of all DepNodes in the graph
-    nodes: IndexVec<SerializedDepNodeIndex, DepNode<K>>,
-    /// The set of all Fingerprints in the graph. Each Fingerprint corresponds to
-    /// the DepNode at the same index in the nodes vector.
-    fingerprints: IndexVec<SerializedDepNodeIndex, Fingerprint>,
-    /// For each DepNode, stores the list of edges originating from that
-    /// DepNode. Encoded as a [start, end) pair indexing into edge_list_data,
-    /// which holds the actual DepNodeIndices of the target nodes.
-    edge_list_indices: IndexVec<SerializedDepNodeIndex, (u32, u32)>,
-    /// A flattened list of all edge targets in the graph. Edge sources are
-    /// implicit in edge_list_indices.
-    edge_list_data: Vec<SerializedDepNodeIndex>,
+    /// The set of all DepNodes in the graph and their position in the mmap.
+    nodes: Option<OwningRef<Mmap, [(DepNode<K>, u32)]>>,
     /// Reciprocal map to `nodes`.
-    index: FxHashMap<DepNode<K>, SerializedDepNodeIndex>,
+    index: RawTable<u32>,
```

Comment:
Before I'll try to understand how hashbrown::RawTable works exactly, should we consider using an [on-disk hash table](https://github.com/rust-lang/odht) here? Then we would not need to build the index during decoding at all. We could also configure it to use a hash function specialized for DepNode.